src/viam/sdk/common/service_helper.hpp

```diff
@@ -0,0 +1,92 @@
+#pragma once
+
+#include <type_traits>
+
+#include <viam/sdk/resource/resource_server_base.hpp>
+
+namespace viam {
+namespace sdk {
+
+class ServiceHelperBase {
+   public:
+    ::grpc::Status fail(::grpc::StatusCode code, const char* message) const noexcept;
+
+    ::grpc::Status failNoRequest() const noexcept;
+
+    ::grpc::Status failNoResource(const std::string& name) const noexcept;
+
+    ::grpc::Status failStdException(const std::exception& xcp) const noexcept;
+
+    ::grpc::Status failUnknownException() const noexcept;
+
+   protected:
+    explicit ServiceHelperBase(const char* method) noexcept : method_{method} {}
+
+   private:
+    const char* method_;
+};
+
+template <typename ServiceType, typename RequestType>
+class ServiceHelper : public ServiceHelperBase {
+   public:
+    ServiceHelper(const char* method, ResourceServer* rs, RequestType* request) noexcept
+        : ServiceHelperBase(method), rs_{rs}, request_{request} {};
+
+    template <typename Callable>
+    ::grpc::Status operator()(Callable&& callable) const noexcept try {
+        if (!request_) {
+            return failNoRequest();
+        }
+        const auto resource = rs_->resource_manager()->resource<ServiceType>(request_->name());
+        if (!resource) {
+            return failNoResource(request_->name());
+        }
+        return invoke_(std::forward<Callable>(callable), std::move(resource));
+    } catch (const std::exception& xcp) {
+        return failStdException(xcp);
+    } catch (...) {
+        return failUnknownException();
+    }
+
+    auto getExtra() const {
```

Comment:
Another way to go about this would be to use SFINAE again to have the Callable be invoked with `extra`, but only if `RequestType` defines `has_request`. But it felt like the template metaprogramming complexity wasn't worth it.

Reply:
Agree; the helper method is much easier to read given the existing meta programming.